### PR TITLE
[SPARK-43179][SHUFFLE] Allowing apps to control whether their metadata gets saved in the db by the External Shuffle Service

### PR DESCRIPTION
**Pull request by otterc**
### What changes were proposed in this pull request?
This change allows applications to control whether their metadata gets saved in the db. For applications with higher security requirements, storing application secret in the db without any encryption is a potential security risk. While filesystem ACLs can help protect the access to the db, this level of security is not sufficient for some use cases. Such applications can chose to not save their metadata in the db. As a result, these applications may experience more failures in the event of a node restart, but we believe this trade-off is acceptable given the increased security risk.

### Why are the changes needed?
These modifications are necessary to reduce the likelihood of security threats for applications with elevated security requirements.

### Does this PR introduce _any_ user-facing change?
No. Added a configuration `spark.shuffle.server.recovery.disabled` which by default is `false`. When set to `true`, the metadata of the application will not saved in the db.

### How was this patch tested?
Added UTs and also verified with test applications in our test environment.

### CODE REVIEW

**Comment by otterc** [comment]
@mridulm @tgravescs @zhouyejoe @akpatnam25 @shuwang21 Please help review

**Comment by mridulm** [comment]
+CC @attilapiros

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala [hunk 1]
```diff
@@ -128,6 +121,33 @@ private[yarn] class ExecutorRunnable(
...
+        } else {
+          // Authentication is not enabled, so just provide dummy metadata
+          ByteBuffer.allocate(0)
+        }
+      ctx.setServiceData(Collections.singletonMap(serviceName, secretBytes))
+    } else {
+      val payload = new mutable.HashMap[String, Object]()
+      payload.put(SHUFFLE_SERVER_RECOVERY_DISABLED.key, java.lang.Boolean.TRUE)
+      if (secretString != null) {
+        payload.put(ExecutorRunnable.SECRET_KEY, secretString)
```
Comment (zhouyejoe) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala:142]: Do we need to keep the same behavior here, when the secretString is null, we still put ByteBuffer.allocate(0) there in the payload?

Comment (otterc) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala:142]: We don't need to keep the same behavior. If authentication is disabled, we don't need to set the value of the secret key. Also, this was a feedback from @mridulm

**Comment by mridulm** [comment]
Exactly that usecase @tgravescs.
We have usecases where applications require heightened security, and are fine with the recomputation cost due to lost shuffle data.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 2]
```diff
@@ -2507,4 +2507,13 @@ package object config {
...
       .createWithDefault(Int.MaxValue)
+
+  private[spark] val SHUFFLE_SERVER_RECOVERY_DISABLED =
+    ConfigBuilder("spark.shuffle.server.recovery.disabled")
+      .internal()
+      .doc("Set to true for applications that prefer to disable recovery when the External " +
+        "Shuffle Service restarts.")
+      .version("3.5.0")
+      .booleanConf
+      .createWithDefault(false)
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:2518]: This is specific to yarn and not supported by other resource managers - call it out in the description ?
Same for the documentation update below - make it explicit that it is for yarn mode, and make sure it is in the right section in the doc.

File: common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java [hunk 3]
```diff
@@ -407,10 +415,30 @@ private static byte[] dbAppKey(AppId appExecId) throws IOException {
...
-        if (db != null) {
+        if (metaInfo != null) {
+          shuffleSecret = (String) metaInfo.get(SECRET_KEY);
+        } else {
+          shuffleSecret = payload;
+        }
+        if (db != null && AppsWithRecoveryDisabled.isRecoveryEnabledForApp(appId)) {
+          AppId fullId = new AppId(appId);
           byte[] key = dbAppKey(fullId);
           byte[] value = mapper.writeValueAsString(shuffleSecret).getBytes(StandardCharsets.UTF_8);
```
Comment (mridulm) [inline common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java:443]: Can we add a test with `shuffleSecret = null` and `appServiceData` is a json ?

**Review by mridulm — COMMENTED** [review]
Just a couple of comments, rest looks good.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 2 above)
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:2518]: if this is only for yarn should we rename to be spark.yarn.shuffle and only put in yarn docs?

File: common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java [hunk 4]
```diff
@@ -407,10 +415,30 @@ private static byte[] dbAppKey(AppId appExecId) throws IOException {
...
+        logger.warn("Unable to parse application data for service: " + payload);
+        metaInfo = null;
+      }
       if (isAuthenticationEnabled()) {
-        AppId fullId = new AppId(appId);
-        if (db != null) {
+        if (metaInfo != null) {
+          shuffleSecret = (String) metaInfo.get(SECRET_KEY);
+        } else {
+          shuffleSecret = payload;
```
Comment (tgravescs) [inline common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java:438]: this looks like it should be backwards compatible, but please make sure we test with older versions of spark using this newer shuffle service if you haven't already

Comment (otterc) [reply common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java:438]: I tested this with an old client and it didn't see any issues.

File: common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java (hunk 3 above)
Comment (otterc) [reply common/network-yarn/src/main/java/org/apache/spark/network/yarn/YarnShuffleService.java:443]: I fixed the test that I added `executor info of apps should not be stored in db if they want to be excluded. Authentication is turned off` to test this case.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 2 above)
Comment (otterc) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:2518]: - I moved the documentation for this config to `running-on-yarn.md`.
- In the security doc, I moved the definition under `Yarn` section. 
PTAL

**Comment by attilapiros** [comment]
As we can run multiple external shuffle services on the same NM what about introducing just a new config for the whole external shuffle service controlling whether recovery is enabled.

Then those applications with higher security can use the external shuffle service where the recovery is disabled.

I think regarding code complexity and code size it would be a much better solution. WDYT?

**Comment by otterc** [comment]
> As we can run multiple external shuffle services on the same NM what about introducing just a new config for the whole external shuffle service controlling whether recovery is enabled.
> 
> Then those applications with higher security can use the external shuffle service where the recovery is disabled.

It can be done but setting up and running  another shuffle service requires much more effort at least in our production environment. 

> I think regarding code complexity and code size it would be a much better solution. WDYT?

The code change here is quite small. Agreed that disabling recovery for the whole ESS may need much smaller change but there is not much complexity with this change as well.

**Comment by tgravescs** [comment]
lgtm

**Review by mridulm — APPROVED** [review]

**Comment by mridulm** [comment]
Thanks for fixing this @otterc !
Thanks for the reviews @tgravescs, @zhouyejoe :-)

**Pull request closed without merging**